src/EditorFeatures/Core/CommandHandlers/GoToNextAndPreviousMethodCommandHandler.cs

```diff
@@ -0,0 +1,185 @@
+            if (document == null || !document.SupportsSyntaxTree || !documentSupportsSuggestionService.SupportsNavigationToAnyPosition(document))
+            {
+                return false;
+            }
+
+            if (!caretPoint.HasValue)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private void ExecuteCommand(Action nextHandler, ITextBuffer subjectBuffer, ITextView textView, bool next)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var caretPoint = textView.GetCaretPoint(subjectBuffer);
+            if (!IsAvailable(document, caretPoint))
+            {
+                nextHandler();
+                return;
+            }
+
+            int? targetPosition = null;
+            var waitResult = _waitIndicator.Wait(EditorFeaturesResources.Navigating, allowCancel: true, action: waitContext =>
+            {
+                targetPosition = GetTargetPosition(document, caretPoint.Value.Position, next, waitContext.CancellationToken);
+            });
+
+            if (waitResult == WaitIndicatorResult.Canceled || targetPosition == null)
+            {
+                return;
+            }
+
+            textView.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(subjectBuffer.CurrentSnapshot, targetPosition.Value), _outliningManagerService);
+        }
+
+        /// <summary>
+        /// Internal for testing purposes.
+        /// </summary>
+        internal static int? GetTargetPosition(Document document, int caretPosition, bool next, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFactsService == null)
+            {
+                return null;
+            }
+
+            var root = document.GetSyntaxRootAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+            var members = syntaxFactsService.GetMethodLevelMembers(root);
+
+            if (members.Count == 0)
+            {
+                return null;
+            }
+
+            var currentMember = GetMember(syntaxFactsService, caretPosition, root);
+            int indexOfCurrentMember;
+            if (currentMember == null)
+            {
+                // We're not contained in any member currently.  Find the first member to start after our
+                // position and use that.
+                for (indexOfCurrentMember = 0; indexOfCurrentMember < members.Count; indexOfCurrentMember++)
+                {
+                    if (members[indexOfCurrentMember].FullSpan.Start > caretPosition)
+                    {
+                        break;
+                    }
+                }
+
+                if (indexOfCurrentMember == members.Count)
+                {
+                    // After the last member, count ourselves as part of that last member, so that we wrap
```

Comment:
Do you mean you're before all members? If you're after all members, that means you should have taken the break early.